components/script/dom/xpathevaluator.rs

```diff
@@ -0,0 +1,101 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::rc::Rc;
+
+use dom_struct::dom_struct;
+use js::rust::HandleObject;
+
+use super::bindings::error::Error;
+use crate::dom::bindings::codegen::Bindings::XPathEvaluatorBinding::XPathEvaluatorMethods;
+use crate::dom::bindings::codegen::Bindings::XPathExpressionBinding::XPathExpression_Binding::XPathExpressionMethods;
+use crate::dom::bindings::error::Fallible;
+use crate::dom::bindings::reflector::{reflect_dom_object_with_proto, DomObject, Reflector};
+use crate::dom::bindings::root::{Dom, DomRoot};
+use crate::dom::bindings::str::DOMString;
+use crate::dom::node::Node;
+use crate::dom::window::Window;
+use crate::dom::xpathexpression::XPathExpression;
+use crate::dom::xpathresult::XPathResult;
+use crate::script_runtime::CanGc;
+
+#[dom_struct]
+pub struct XPathEvaluator {
+    reflector_: Reflector,
+    window: Dom<Window>,
+}
+
+impl XPathEvaluator {
+    fn new_inherited(window: &Window) -> XPathEvaluator {
+        XPathEvaluator {
+            reflector_: Reflector::new(),
+            window: Dom::from_ref(window),
+        }
+    }
+
+    pub fn new(
+        window: &Window,
+        proto: Option<HandleObject>,
+        can_gc: CanGc,
+    ) -> DomRoot<XPathEvaluator> {
+        reflect_dom_object_with_proto(
+            Box::new(XPathEvaluator::new_inherited(window)),
+            window,
+            proto,
+            can_gc,
+        )
+    }
+}
+
+impl XPathEvaluatorMethods<crate::DomTypeHolder> for XPathEvaluator {
+    fn Constructor(
+        window: &Window,
+        proto: Option<HandleObject>,
+        can_gc: CanGc,
+    ) -> DomRoot<XPathEvaluator> {
+        XPathEvaluator::new(window, proto, can_gc)
+    }
+
+    fn CreateExpression(
+        &self,
+        expression: DOMString,
+        _resolver: Option<
+            Rc<crate::dom::bindings::codegen::Bindings::XPathNSResolverBinding::XPathNSResolver>,
+        >,
+    ) -> DomRoot<XPathExpression> {
+        let global = self.global();
+        let window = global.as_window();
+        // NB: this function is *not* Fallible according to the spec, so we swallow any parsing errors and
+        // just pass a None as the expression... it's not great.
+        let parsed_expression = crate::xpath::parse(&expression);
+        XPathExpression::new(window, None, CanGc::note(), parsed_expression.ok())
```

Comment:
This irks me. The spec's IDL says `[NewObject] XPathExpression createExpression(...)` and apparently I can't put both `[NewObject]` and `[Throws]` there... maybe I just don't know the syntax.

Reply:
its `[NewObject, Throws]`

Reply:
Can't believe I didn't think to try that... thanks!